### ISSUE TITLE
SC2048 vs SC2066 - loop over array .. use "quotes" vs "the loop will only run once"

### ISSUE THREAD

**Issue by PeterPitterling** [koalaman/shellcheck#2374]
#### For bugs
- Rule Id : SC2048 , SC2066
- My shellcheck version (`shellcheck --version` or "online"):  online
- [x] The rule's wiki page does not already cover this (e.g. https://shellcheck.net/wiki/SC2048)
- [x] I tried on https://www.shellcheck.net/ and verified that this is still a problem on the latest commit

#### For new checks and feature suggestions
- [x] https://www.shellcheck.net/ (i.e. the latest commit) currently gives no useful warnings about this
- [x] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

#### Here's a snippet or screenshot that shows the problem:

```sh
#!/usr/bin/env bash

#SC2048
for item in ${array[*]:-}; do

 #do dummy stuff
 item=${item##*/}

done

#SC2066
for item in "${array[*]:-}"; do

 #do dummy stuff
 item=${item##*/}

done
```

#### Here's what shellcheck currently says:
$ shellcheck myscript
 
Line 4:
for item in ${array[*]:-}; do
            ^-- SC2048 (warning): Use "${array[@]}" (with quotes) to prevent whitespace problems.
 
Line 12:
for item in "${array[*]:-}"; do
            ^-- SC2066 (error): Since you double quoted this, it will not word split, and the loop will only run once.

#### Here's what I wanted or expected to see:
It is more a question, how to deal with it? Just running without quotes and suppressing shellcheck sc2048?
Couldn't be sc a bit smarter here :) ?

**Comment by koalaman** [comment]
I must be missing something. Why are you not using `"${array[@]}"` as originally suggested?

**Comment by PeterPitterling** [comment]
because it was not working for me .. but i found why ..
I had some code (non-touched for long time) like this:

```shell
local -r IFS=$'\n'; platform=( "$(lscpu)" )

for line in ${platform[*]:-}; do
 :
 #...
done
```

i quoted the command to avoid SC2207 "Prefer mapfile or read -a to split command output (or quote to avoid splitting)."
This worked for me i guess because of IFS ... but in the moment i use quotes of course not ...

I changed this now like suggested in SC2207

```shell
    if [[ ${BASH_VERSINFO[0]:-} -ge 4 ]] ; then
        mapfile -t platform < <(lscpu)
    else
        platform=()
        while IFS='' read -r line; do platform+=("$line"); done < <(lscpu)
    fi

    local line
    for line in "${platform[@]:-}"; do
      :
     #...
   done
```

**Comment by PeterPitterling** [comment]
here is a pitfall with empty arrays and set -u and recommendation from SC2048 "${array[@]}" or "${array[@]:-}" .. the behavior changes because with quotes it will run at least 1 time now.

```shell 
#!/usr/bin/env bash
set -u

declare -a array=()

#for item in ${array[@]}; do
# printf "throw error - unbound variable \n\n"
#done

for item in ${array[@]:-}; do
 printf "will not run \n\n"
done

for item in "${array[@]:-}"; do
 printf "will run at least ONE time\n\n"         #not intended !!
done

for item in ${array[@]+"${array[@]}"}; do
 printf "would be correct\n\n"
done

```

shouldn't that be Shellchecks recommendation for arrays ?
